op-chain-ops/cmd/celo-migrate/state.go

```diff
@@ -242,48 +246,70 @@ func applyStateMigrationChanges(config *genesis.DeployConfig, genesis *core.Gene
 // If an account already exists, it adds the balance of the new account to the existing balance.
 // If the code of an existing account is different from the code in the genesis block, it logs a warning.
 // This changes the state root, so `Commit` needs to be called after this function.
-func applyAllocsToState(db *state.StateDB, genesis *core.Genesis, config *params.ChainConfig) {
+func applyAllocsToState(db vm.StateDB, genesis *core.Genesis, config *params.ChainConfig) error {
 	log.Info("Starting to migrate OP contracts into state DB")
 
-	accountCounter := 0
+	copyCounter := 0
+	skipCounter := 0
 	overwriteCounter := 0
-	for k, v := range genesis.Alloc {
-		accountCounter++
+	whitelist := accountOverwriteWhitelist[config.ChainID.Uint64()]
 
-		balance := uint256.MustFromBig(v.Balance)
+	for k, v := range genesis.Alloc {
```

Comment:
This is getting quite complicated and I don't think it needs to be, I think the following for loop would be sufficient.

```
	for k, v := range genesis.Alloc {
		// Check that the balance of the account to written is zero,
		// as we must not create new CELo tokens
		if v.Balance.Cmp(big.NewInt(0)) != 0 {
			log.Error("Account balance is not zero, would change celo supply", "address", k.Hex())
			return fmt.Errorf("account balance is not zero, would change celo supply: %s", k.Hex())
		}

		if db.Exist(k) {
			_, whitelisted := whitelist[k]

			// If the account is not whitelisted and has a non zero nonce or code size, bail out we will need to manually investigate how to handle this.
			if !whitelisted && (db.GetCodeSize(k) > 0 || db.GetNonce(k) > 0) {
				return fmt.Errorf("account exists and is not whitelisted, account: %s, nonce: %d, codeSize: %d", k.Hex(), db.GetNonce(k), db.GetCodeSize(k))
			}
			overwriteCounter++
		}
		copyAlloc(db, k, v, v.Nonce)
		copyCounter++
	}
```

Reply:
This way we loose the specific errors, but that might be fine given that we fail anyways.

Reply:
I didn't feel that the specific errors were that helpful, they were all variations of (we're failing because there is an existing account) we could extend the error message in the code I posted above to include whether the contract code matches and then I think all the information that was included in all the specific errors should be covered.

>But this approach also limits our options when we would decide to keep the existing nonce in any of the cases we discussed above.

That is true and thinking about it more there doesn't seem to be any clear or consistent way to decide if the nonce should be kept or overwritten. I think we have to asses this on a case by case basis so the decision to keep or not keep the nonce should go into the whitelist.

In the case of the create2 deployer, it seems as if the nonce value is not particularly important, since the nonce is used to derive the contract address when calling the create method but not when calling the create2 method, and given the name of the create2 deployer it sounds like it's not going to be calling create ( I haven't actually checked). So I think we could use either. But in cases where contracts are calling create we'll need to ensure that we keep the existing nonce.

Reply:
Done

Reply:
The whitelist is still lacking a decision about whether to keep the nonce. You could just make it a `map[common.Address]bool` and keep the nonce if the bool is true.

Reply:
Implemented in [77fbec6](https://github.com/celo-org/optimism/pull/202/commits/77fbec69028e3b92c2b2d09356d77a8f7663bb31)